### PR TITLE
Fix new result pane column behavior

### LINKED ISSUE #549: Default column for a new results pane
MSSQL Extension Version: 0.2.0
VSCode Version: 1.8
OS Version: Win10

Steps to Reproduce:
1. Open three columns in VSCode. 
2. Run a query that opens a results pane.

Actual: 
Results pane opens in 2nd column (current default for new panes). 

Expected:
Not sure. Where should the results pane default to?

### PR DESCRIPTION
**Pull request by Raymondd**
Fixes #549.
splitPaneSelection configuration property added. Corresponding behavior added to vscodeWrapper class.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -441,8 +441,17 @@
           "type": "boolean",
           "description": "[Optional] Configuration options for copying multi-line results from the Results View",
           "default": true
+        },
+        "mssql.splitPaneSelection": {
+          "type": "string",
+          "description": "[Optional] Configuration options for which column new result panes should open in (Options: {'next'|'same'|'last'})",
```
Comment (kevcunnane) [inline package.json]: I think you can remove the `(Options: {'next'|'same'|'last'})` part - the enum values are listed as options by VSCode when altering them

File: package.json [hunk 2]
```diff
@@ -441,8 +441,17 @@
...
+          "default": "next",
+          "enum": [
+            "next",
+            "same",
```
Comment (kevcunnane) [inline package.json]: `current` instead of `same`, `end` instead of `last` unless you mean `previous` ?

File: src/controllers/vscodeWrapper.ts [hunk 3]
```diff
@@ -269,4 +269,5 @@ export default class VscodeWrapper {
     public get workspaceRootPath(): string {
         return vscode.workspace.rootPath;
     }
+
```
Comment (kevcunnane) [inline src/controllers/vscodeWrapper.ts]: Seems like you can undo the change in this file

File: src/models/SqlOutputContentProvider.ts [hunk 4]
```diff
@@ -313,16 +313,15 @@ export class SqlOutputContentProvider implements vscode.TextDocumentContentProvi
         let paneTitle = Utils.formatString(Constants.titleResultsPane, queryRunner.title);
         // Always run this command even if just updating to avoid a bug - tfs 8686842
 
-        // Find any URIs which match the one we are about to display
-        let resultPaneURIMatch = vscode.workspace.textDocuments.find(tDoc => tDoc.uri.toString() === resultsUri);
 
         // Check if the results window already exists
-        if (resultPaneURIMatch !== undefined) {
+        if (this.doesResultPaneExist(resultsUri)) {
             // Implicity Use existsing results window by not providing an pane
```
Comment (kevcunnane) [inline src/models/SqlOutputContentProvider.ts]: typo `existing` instead of `existsing` (since you're changing this area anyhow)

File: src/models/SqlOutputContentProvider.ts [hunk 5]
```diff
@@ -535,4 +534,51 @@ export class SqlOutputContentProvider implements vscode.TextDocumentContentProvi
         }
         return path.join(os.tmpdir(), columnName + '_' + String(Math.floor( Date.now() / 1000)) + String(process.pid) + '.' + linkType);
     }
+
+    /**
+     * Returns wether or not a result pane with the same URI exists
```
Comment (kevcunnane) [inline src/models/SqlOutputContentProvider.ts]: typo `whether`

File: src/models/SqlOutputContentProvider.ts [hunk 6]
```diff
@@ -535,4 +534,51 @@ export class SqlOutputContentProvider implements vscode.TextDocumentContentProvi
...
+     * @param The string value of a Uri.
+     * @return Existence boolean
```
Comment (kevcunnane) [inline src/models/SqlOutputContentProvider.ts]: `@return boolean true if pane exists`

File: src/models/SqlOutputContentProvider.ts [hunk 7]
```diff
@@ -535,4 +534,51 @@ export class SqlOutputContentProvider implements vscode.TextDocumentContentProvi
...
+     * public for testing purposes
+     */
+    public doesResultPaneExist(resultsUri: string): boolean {
+        let resultPaneURIMatch = vscode.workspace.textDocuments.find(tDoc => tDoc.uri.toString() === resultsUri);
```
Comment (kevcunnane) [inline src/models/SqlOutputContentProvider.ts]: please use `this._vscodeWrapper.textDocuments()` instead of direct references so it's easier to mock.

File: src/models/SqlOutputContentProvider.ts [hunk 8]
```diff
@@ -535,4 +534,51 @@ export class SqlOutputContentProvider implements vscode.TextDocumentContentProvi
...
+        let splitPaneSelection = config[Constants.configSplitPaneSelection];
+        let viewColumn: vscode.ViewColumn;
+
+
+        switch (splitPaneSelection) {
+        case 'same' :
+            viewColumn = this._vscodeWrapper.activeTextEditor.viewColumn;
+            break;
+        case 'last' :
+            viewColumn = vscode.ViewColumn.Three;
```
Comment (kevcunnane) [inline src/models/SqlOutputContentProvider.ts]: Is three the max value? Could this ever change?

File: test/sqlOutputContentProvider.test.ts [hunk 9]
```diff
@@ -1,105 +1,160 @@
...
+            done();
+        } catch (err) {
+            done(new Error(err));
+        }
+    });
+
+});
+
+
+// TODO: rewrite all the outputprodiver handle tests (old ones kept for reference)
```
Comment (kevcunnane) [inline test/sqlOutputContentProvider.test.ts]: is there an issue tracking it? if so what's the issue #? also typo `outputprovider`

File: test/sqlOutputContentProvider.test.ts [hunk 10]
```diff
@@ -1,105 +1,160 @@
...
+            });
+        }
+
+        class Case {
+            position: number;
+            config: string;
+            expectedColumn: number;
+        }
+
+        let cases: Case[] = [
```
Comment (kevcunnane) [inline test/sqlOutputContentProvider.test.ts]: It would be nice if each of these was a test case. See https://mochajs.org/#dynamically-generating-tests for how to easily convert this (basically define cases at the `suite` level and in the foreach define a test.

File: package.json (hunk 2 above)
Comment (Raymondd) [inline package.json]: It does mean 'end'. Will change.

File: src/models/SqlOutputContentProvider.ts (hunk 8 above)
Comment (Raymondd) [inline src/models/SqlOutputContentProvider.ts]: Three is the maximum number of panes one can have open in VSCode.

File: test/sqlOutputContentProvider.test.ts (hunk 10 above)
Comment (Raymondd) [inline test/sqlOutputContentProvider.test.ts]: I could dynamically generate each test, but would it be worth it to have 9 test cases and an entire suite for this one function?

Comment (kevcunnane) [inline test/sqlOutputContentProvider.test.ts]: Sounds good
